### PR TITLE
Inter-PodAffinity is calculated on multiple pods

### LINKED ISSUE #68701: Support inter-Pod affinity to one or more Pods
In the current implementation of inter-Pod affinity, the scheduler looks for a single existing pod that can satisfy all the terms of inter-pod affinity of an incoming pod.
With the recent changes made to the implementation of inter-Pod affinity, we can now support multiple pods satisfying inter-pod affinity. One of the main reasons we didn't pursue the idea before was the fact that the inter-pod affinity feature was very slow (3 orders of magnitude slower than other scheduler predicates). We didn't want to add more complication to an already slow predicate. However, we can now think about adding the feature.
With this feature, a pod can have multiple affinity terms satisfied by a group of pods, as opposed to only a single pod. For example:

```
- Assume that the cluster has two nodes: 
- nodeA located in in zone1/region1
- nodeB located in zone2/region1

- There are two existing pods on these nodes:
- Pod1:
    - nodeName: "nodeA"
    - label "foo":""

- Pod2:
    - nodeName: "nodeB"
    - label "bar":""

- Pod3 comes in with inter-pod affinity:
    - affinity terms:
        - {label "foo" exists, topologyKey: "region"}
        - {label "bar" exists, topologyKey: "zone"}
```

With our current (K8s 1.12) implementation, Pod3 is not schedulable, because there is no single pod that satisfies all of its affinity terms. However, if we support multiple pods satisfying the affinity terms, Pod3 can be scheduled on nodeB. Pod1 satisfies the first term of its affinity in region1 and Pod2 satisfies its second term in zone2. So, any node in zone2/region1 will be feasible for Pod3.

Given the current implementation of inter-pod affinity and the use of "Topology Pair Maps", I believe implementing this feature requires little changes and won't have noticeable performance impact.

/kind feature
/sig scheduling

cc/ @Huang-Wei @ahmad-diaa

### PR DESCRIPTION
**Pull request by Huang-Wei**
**What this PR does / why we need it**:
Inter-PodAffinity is now calculated on multiple pods instead of single pod.

**Which issue(s) this PR fixes**:
Fixes #68701

**Special notes for your reviewer**:

**Release note**:
```release-note
Inter-PodAffinity is now calculated on multiple pods instead of single pod.
```

/sig scheduling
/kind feature

### CODE REVIEW

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 1]
```diff
@@ -1334,13 +1334,27 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
...
+			return false
+		}
+
+		pair := topologyPair{key: term.TopologyKey, value: topologyValue}
+		podSet, ok := topologyPairs.topologyPairToPods[pair]
+		if !ok {
+			return false
+		}
+		// "idx" should be found in "podSet[<any pod>]"
+		found := false
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/predicates.go]: the `found` variable/logic could be omitted by using a continue <outer loop> trick.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 2]
```diff
@@ -1334,15 +1334,26 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
...
 			return false
 		}
+
+		pair := topologyPair{key: term.TopologyKey, value: topologyValue}
+		podSet, ok := topologyPairs.topologyPairToPods[pair]
+		if !ok {
+			return false
+		}
+		// "idx" should be found in "podSet[<any pod>]"
+		for _, bitwisedIdx := range podSet {
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Why do we need this check? I appears to me that when all topologyPairs of terms exist in the topologyPairs.topologyPairToPods, affinity is satisfied.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: It's because right now `topologyPairs.topologyPairToPods[pair]`, which is `podSet map[*v1.Pod]uint64`, maintain a relation of "which pod matches on which index of the terms" - it's like a "contribution" to the final match on all terms of PodAffinity. Additionally, two pods in different nodes can "contribute" to the same term, and failed on other term(s). So counting numbers doesn't help, here I used the bitwised ORed value.

So in this method, the logic is: only when all terms's index can be found in some pod, we treat current node as a match.

("some pod" is not necessarily on current node, but it's in the same topology domain of the node)

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: @bsalamat latest commit 2e88364e34157201d75b7931e96ca6d7964d6c14 added a testcase which failed if we don't have bitmap check.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: But that case can be resolved by "merging" mergeable terms (terms which has the same topologyKey).

Let me think more to see if there are any other corner cases.

Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > @bsalamat latest commit 2e88364 added a testcase which failed if we don't have bitmap check.

Yes, you are right. In that case, we should replace the bit map with an actual map to store the indeces of terms. I understand that probably no one adds more than 64 terms to an affinity rule, but we cannot have that restriction in our implementation.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 3]
```diff
@@ -1334,15 +1334,26 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
...
+			// this indicates current "idx" is satisfied
+			if bitwisedIdx|(1<<uint(idx)) == bitwisedIdx {
+				continue OUTERLOOP
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: Why don't you just `break` here? Those labels remind me of `goto`.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 2 above)
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > but we cannot have that restriction in our implementation

Understood.

Basically here is what I'm thinking of at this point:

- option1: continue current solution, but need extra efforts on server validation (maybe as well as kubectl validation), and also docs
- option2: don't use bitmap, revert back to a simple struct{}. And when computing topologyPairToPods for PodAffinity, perform a "dry-run merge" on terms which has the same topology. Say only when a "merged" term matches, we add it to topologyPairToPods.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 3 above)
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: I see. I would add flags to cover that. Please don't use labels.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 2 above)
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: option3: use an actual map/set to store the indexes of matched terms instead of the current bitmap. ;-)

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: sure, option3 would be the most reasonable one.

I can make podSet from current `map[*v1.Pod]struct{}` to `map[*v1.Pod]intSet` (`type intSet map[int]struct{}`).

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 3 above)
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: sure, I actually used https://github.com/kubernetes/kubernetes/pull/68725/commits/20b71555fba4520df7c0e9c2b2f6714c8f4678e5 to "try out" the `continue LABEL` directive. Before that, I used a flag to achieve same thing.

File: pkg/scheduler/algorithm/predicates/predicates.go (hunk 2 above)
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Yes, that would be better IMO.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 4]
```diff
@@ -1334,13 +1334,27 @@ func (c *PodAffinityChecker) satisfiesExistingPodsAntiAffinity(pod *v1.Pod, meta
 //  topology of all the "terms" for the given "pod".
 func (c *PodAffinityChecker) nodeMatchesAllTopologyTerms(pod *v1.Pod, topologyPairs *topologyPairsMaps, nodeInfo *schedulercache.NodeInfo, terms []v1.PodAffinityTerm) bool {
```
Comment (ahmad-diaa) [inline pkg/scheduler/algorithm/predicates/predicates.go]: @bsalamat @Huang-Wei I think ``pod *v1.Pod`` can be removed from the parameters as it's not used anymore.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: @ahmad-diaa Thanks for the catch.

**Comment by Huang-Wei** [comment]
Removed "[WIP]" from title.

@bsalamat @ahmad-diaa code is ready for review.

File: pkg/scheduler/algorithm/predicates/predicates_test.go [hunk 5]
```diff
@@ -3982,16 +3983,83 @@ func TestInterPodAffinityWithMultipleNodes(t *testing.T) {
...
 				{ErrPodAffinityNotMatch, ErrPodAffinityRulesNotMatch},
+				{ErrPodAffinityNotMatch, ErrPodAffinityRulesNotMatch},
 			},
 			fits: map[string]bool{
 				"nodeA": false,
 				"nodeB": false,
+				"nodeC": false,
 			},
-			name: "Test incoming pod's affinity: firstly check if all affinityTerms matches, and then check if all topologyKeys match, and the match logic should be satified on the same pod",
+			name: "Test incoming pod's affinity: pod1 and pod2 match partially on the same toplogy domain (zone), so they are not fits",
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates_test.go]: s/toplogy/topology/

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 6]
```diff
@@ -1374,13 +1388,13 @@ func (c *PodAffinityChecker) satisfiesPodsAffinityAntiAffinity(pod *v1.Pod,
...
 		if affinityTerms := GetPodAffinityTerms(affinity.PodAffinity); len(affinityTerms) > 0 {
-			matchExists := c.nodeMatchesAllTopologyTerms(pod, topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
+			matchExists := c.nodeMatchesAllTopologyTerms(topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
 			if !matchExists {
 				// This pod may the first pod in a series that have affinity to themselves. In order
 				// to not leave such pods in pending state forever, we check that if no other pod
 				// in the cluster matches the namespace and selector of this pod and the pod matches
 				// its own terms, then we allow the pod to pass the affinity check.
-				if !(len(topologyPairsPotentialAffinityPods.topologyPairToPods) == 0 && targetPodMatchesAffinityOfPod(pod, pod)) {
+				if !targetPodMatchesAffinityOfPod(pod, pod) {
```
Comment (bsalamat) [inline pkg/scheduler/algorithm/predicates/predicates.go]: This is now incorrect. In the previous code we were checking that when there was no pod in the whole cluster that matched the affinity of the incoming pod AND the incoming pod didn't match its own affinity, we would return "NotMatch". This PR has eliminated the first part of the condition. So, if the incoming pod's affinity is not satisfied on this node, we will not return "NotMatch" when the pod matches its own affinity, regardless of the fact that there might be other pods in the cluster that match the incoming pod's affinity.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: So is that to say, currently it needs to get an "aggregated" match result in advance which represents "if there is some node matches incoming pod's affinity".

- If the "aggregated" match result is true, then just give final fit judgement depending on `matchExists`
- If the "aggregated" match result is false, and `matchExists` is also false, then proceed to check if its podAffinity matches its own labels.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > it needs to get an "aggregated" match result in advance which represents "if there is some node matches incoming pod's affinity".

This isn't that easy to implement based on current code.

Firstly, we can't simply do a pre-compute b/c for some of nodes, predicates would have failed before going to InterPodAffinity predicate. So an arbitrary pre-compute is impractical and inefficient.

Secondly, like what current logic is, it's computed on demand. But it should be able to achieve that if `!matchExists`, hold on until it gets a signal indicating either 1) some other node passed PodAffinity check, or 2) nodes running InterPodAffinity predicate (note a: not all nodes) have all failed on PodAffinity check.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > Secondly, like what current logic is, it's computed on demand. But it should be able to achieve that if `!matchExists`, hold on until it gets a signal indicating either 1) some other node passed PodAffinity check, or 2) nodes running InterPodAffinity predicate (note a: not all nodes) have all failed on PodAffinity check.

Current predicates framework assumes predicates can be run independently on each node. So it's pretty challenging to have above logic implemented in InterPodAffinity predicate.

Another implementation is to do a post-handling after all predicates finish. But this requires (1) InterPodAffinity predicate is the last predicate (it's the case right now) and (2) inside InterPodAffinity predicate, PodAffinity check should also be handled at last step (specifically, in satisfiesPodsAffinityAntiAffinity(), the logic of handling affinity should be moved after logic of handling antiAffinity).

A quick implementation is [here](https://gist.github.com/Huang-Wei/87e4142709dfc204f7621e8a7b525edb).

Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > @bsalamat I believe case https://gist.github.com/Huang-Wei/008dae22f0bba635ed4f1ad151f9768c is what you meant.

Yes, except that the pod is schedulable on both nodes. It has affinity to itself. Since its affinity doesn't match any pods in the cluster, but itself, its affinity become a no-op and it can be scheduled on any node as long as it passes other predicates.

> Current predicates framework assumes predicates can be run independently on each node. So it's pretty challenging to have above logic implemented in InterPodAffinity predicate.

Yes. I agree. I don't have any good solution in mind at the moment.

> Another implementation is to do a post-handling after all predicates finish. But this requires (1) InterPodAffinity predicate is the last predicate (it's the case right now) and (2) inside InterPodAffinity predicate, PodAffinity check should also be handled at last step (specifically, in satisfiesPodsAffinityAntiAffinity(), the logic of handling affinity should be moved after logic of handling antiAffinity).

Yes, but that's very brittle. These assumptions may be valid in the current implementation, but they could change in the future. We shouldn't corner ourselves by making these assumptions invariant.

Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: Thanks @bsalamat. Those are almost the same understanding from me. Let me think more to give a more elegant solution.

BTW: regarding

> Yes, except that the pod is schedulable on both nodes.

Maybe you mis-read that case - nodeA is a fit actually :) So only nodeA is a fit.

Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/predicates.go]: > Maybe you mis-read that case - nodeA is a fit actually :) So only nodeA is a fit.

Yes, I didn't notice that. You are right. NodeA is the only one.

**Comment by Huang-Wei** [comment]
@bsalamat Some changes since your last review:

1. `podAffinityMatchMap map[int]int` is introduced into `topologyPairsMaps`. It represents a mapping on which index of podAffinityTerms (key) is matched on how many pods (value).
    - it's updated on initilization, meta.AddPod and meta.RemovePod
    - initilization is performed on all nodes in parrellel, but b/c there is already a mutex in `appendResult`. So unnecessary to use `atomic.AddInt32` during doing math on `podAffinityMatchMap`.
    - for podAntiAffinity related topologyMaps, `podAffinityMatchMap map[int]int` holds a "zero value"
1. a bool variable `isPodAffinitySatisfied` is introduced in `predicateMetadata`, which can be used externally to quickly check if a pod's podAffinity can fit in the whole cluster. BTW: it might be helpful also in preemption logic.
    - `isPodAffinitySatisfied` is updated in meta initilization, meta.AddPod and meta.RemovePod when necessary
1. replace calling of `targetPodMatchesAffinityOfPod` to a new `podMatchesItsOwnAffinityOnNode` as matching its own podAffinity is not enough, we should also consider if target node has proper topology labels.
1. adding some extra testcases

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 7]
```diff
@@ -1374,13 +1388,13 @@ func (c *PodAffinityChecker) satisfiesPodsAffinityAntiAffinity(pod *v1.Pod,
...
-			matchExists := c.nodeMatchesAllTopologyTerms(pod, topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
+			matchExists := c.nodeMatchesAllTopologyTerms(topologyPairsPotentialAffinityPods, nodeInfo, affinityTerms)
 			if !matchExists {
 				// This pod may the first pod in a series that have affinity to themselves. In order
-				// to not leave such pods in pending state forever, we check that if no other pod
+				// to not leave such pods in pending state forever, we check that if no other node
 				// in the cluster matches the namespace and selector of this pod and the pod matches
 				// its own terms, then we allow the pod to pass the affinity check.
-				if !(len(topologyPairsPotentialAffinityPods.topologyPairToPods) == 0 && targetPodMatchesAffinityOfPod(pod, pod)) {
+				if predicateMeta.isPodAffinitySatisfied || !podMatchesItsOwnAffinityOnNode(pod, node) {
```
Comment (Huang-Wei) [inline pkg/scheduler/algorithm/predicates/predicates.go]: seems moving this check to line 1388 makes more sense. If so the logic'd be:

- if incomingPod's podAffinity is satisfied, but we have no idea which nodes are fits, so continue to see if current node is a fit
- if incomingPod's podAffinity is not satisfied, then for sure it's not a fit on current node, so don't need to calculate `matchExists`; instead, directly check if `podMatchesItsOwnAffinityOnNode`

cc @bsalamat for brainstorming

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 8]
```diff
@@ -227,6 +259,49 @@ func (meta *predicateMetadata) RemovePod(deletedPod *v1.Pod) error {
...
+		}
+	}
+	return nil
+}
+
+// satifiesPodAffinity calculates if podAffinity of "pod" can fit or not
+// on the whole cluster
+// note: this function is not expected to be called externally (outside metadata.go)
+// other potential callers should use meta.isPodAffinitySatisfied instead
+func satifiesPodAffinity(pod *v1.Pod, podAffinityMatchMap map[int]int) bool {
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: @bsalamat you're right, `map[int]int` is not enough. Let me update that data structure.

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: > In fact figuring out whether the node affinity can be satisfied anywhere in the cluster without actually checking all the nodes is the biggest challenge.

Agree. My current idea is to first implement a helper structure which can return all node names by searching topologyKey (a.k.a node label) in O(1) time. It's quite similar like `NodeTree`.

Then based on this, I can know which nodes fit for which affinityTerm, and doing an intersection among nodes would give the final result.
The time complexity would be O(n•log<sub>2</sub>m)
- n means length of node matching a specific affinityTerm
- m means length of affinityTerms.

But of course we need O(n•m) space.

@bsalamat do you think it's a viable approach?

[the remark this answers is not in the document]
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/metadata.go]: At this point, I feel this feature is not worth the complexity given that we have not seen any strong desire from users to have it. When we have the scheduling framework, this can be easily implemented as a post-filter plugin. Once all filters (predicates) are executed, we check:
1. Pod is unschedulable.
2. Pod has affinity to itself.
3. Find nodes that have failed "inter-pod affinity" only.
4. Mark those nodes feasible for the pod.

[the remark this answers is not in the document]
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/metadata.go]: Well, the above algorithm is correct, only if we run all predicates for each node even after some of the predicates have failed.

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: > When we have the scheduling framework, this can be easily implemented as a post-filter plugin.

That makes more sense - it's not a good deal to make it as default behavior with such cost.

> Well, the above algorithm is correct, only if we run all predicates for each node even after some of the predicates have failed.

Yes, it's not a "on-demand calculation" algorithm.

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: So @bsalamat are you ok with proceeding with (1) discard recent 2 commits which are for the special PodAffinity-match-itself case, (2) remove check `targetPodMatchesAffinityOfPod(pod, pod))` as the behavior is already incorrect

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: Thanks I got it. Let me do a summary here:

Right now, in terms of supporting PodAffinity on multiple nodes, current PR implements most cases. So we can keep relative commits as is.

Regarding the case support podAffinity to self. Here is current status and possible way to proceed without new scheduling framework:

1. I'm in halfway of implementing the idea described in https://github.com/kubernetes/kubernetes/pull/68725#discussion_r222424295. And as you pointed out, it's an algorithm happening in calculating/updating PredicateMetadata phase. In other words, it's calculated against all nodes in advance.

2. Another way is to use chan/sync.cond in InterPodPredicate phase. Say to only calculate against nodes which have passed previous predicates.

@bsalamat for solution 1, by considering the time/space complexity and behavior, I think your take is to stop investing on it, right?

If so, based on fact that scheduling framework won't come too soon, I'd give a try on solution 2, but will lower the priority in my list. Sounds good?

[the remark this answers is not in the document]
Comment (bsalamat) [reply pkg/scheduler/algorithm/predicates/metadata.go]: Yes to postponing solution 1, but your solution 2 is not clear to me. Can you elaborate on it? Please keep in mind that matching namespace and selector of inter-pod affinity should be done for all pods in a cluster, not only for pods on nodes that have passed previous predicates.

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: > your solution 2 is not clear to me. Can you elaborate on it?

Right now it's a rough idea - pass in a shared chan/sync.cond into PodAffinityChecker, so that for a check (on one node), it can communicate with other checks: broadcast() if it failed on podAffinity check specifically, or close the chan if it passed, so that other failed checkers don't need to wait for a (possible) eventual all-failure signal.

> Please keep in mind that matching namespace and selector of inter-pod affinity should be done for all pods in a cluster

Sure, topologyKeyMaps are calculated/synced in PredicateMetadata phase, and they are always calculated against all pods. It's still the global thing.

Let me give an example to describe expected behavior in solution 2:

```
- Assume that the cluster has two nodes: 
- nodeA located in in zone1/region1, and has 4 allocatable cpus
- nodeB located in zone2/region1, and has 2 allocatable cpus,

- There are two existing pods on these nodes:
- Pod1:
    - nodeName: "nodeA"
    - no labels
- Pod2:
    - nodeName: "nodeB"
    - label "foo":"", "bar":""

- Pod3 comes in with inter-pod affinity:
    - affinity terms:
        - {label "foo" exists, topologyKey: "region"}
        - {label "bar" exists, topologyKey: "zone"}
    - label "foo":"", "bar":""

- case I: if pod3 request <= 2 cpus, it can fit on nodeB, and cannot fit on nodeA
- case II: if pod3 request > 2 cpus, it cannot fit on nodeB due to not enough cpu in nodeB, and
it can fit on nodeA as pod3 matches itself. Unlike solution 1, for this case whether nodeB can fit is not respected/calculated any more as it failed in previous predicate.
```

[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: Per discussion with @bsalamat in today's sig meeting, solution 2 changes the behavior what we have today. Say in above example, even if pod3 failed on nodeB due to insufficient cpu, nodeB still need to be considered to answer "whether incoming pod can fit on a node".

So I will go back to solution 1, but lower the priority.

**Comment by Huang-Wei** [comment]
@bsalamat based on the benchmark tests introduced in #69898, the performance is not bad (see "after"/"after1" sheets in https://docs.google.com/spreadsheets/d/1GI5mCAO3EAMBqwj3sjQToxrOQCwo9uo_RFIGcw2wiF0/edit?usp=sharing), and testing procedure is described in this [gist](https://gist.github.com/Huang-Wei/f2f34112d908bf8cb69661bea1bbd49b).

**Comment by bsalamat** [comment]
Thanks, @Huang-Wei! I had a quick look at the benchmark results. They are promising and shows no noticeable performance change. I will review the PR soon.

**Comment by bsalamat** [comment]
BTW, it would be great if you run those benchmarks for 5000 node clusters as well. In order to do so, you need to add 5000 node config to the test code.

**Comment by Huang-Wei** [comment]
Seems my 8 core 32 GB machine can't afford... No matter it's with or without this PR, it's able to run through `5000Nodes/250Pods`, but panics every time on `5000Nodes/5000Pods`.

**Comment by Huang-Wei** [comment]
The code can be reviewed now.

Meanwhile I'm compiling a benchmark report with and w/o this PR (for 5000 nodes). Will share later.

**Comment by Huang-Wei** [comment]
Benchmark test: https://docs.google.com/spreadsheets/d/15w-Ixz9zRjpVPUMXlED2nrUNOsM4su_YxF7kln28zG8/edit?usp=sharing

**Comment by Huang-Wei** [comment]
@misterikkit I wrote a shell script to collect the testing results: https://gist.github.com/Huang-Wei/a17dce0e4cf924e8af481f307bfec6ad.

But I haven't automated the spreadsheet part - I copied them out to the speadsheet for now :)

**Comment by Huang-Wei** [comment]
@bsalamat @misterikkit do you have any further comments on this PR?

File: pkg/scheduler/nodeinfo/topology_info.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
+/*
```
Comment (bsalamat) [inline pkg/scheduler/nodeinfo/topology_info.go]: The question is whether we should place topology_info.go and its tests in the "internal" directory or here. The "nodeinfo" directory is not in the internal directory, because other modules use it. We should put topology_info under nodeinfo if we intend to allow other modules (autoscaler, node, etc) to use it and we want to provide backward compatibility. Otherwise, it shouldn't be here.

File: pkg/scheduler/nodeinfo/topology_info.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+	"k8s.io/kubernetes/pkg/apis/core/helper"
+)
+
+// TopologyPair is a key/value pair, and now used to describe labels of a node
+type TopologyPair struct {
+	Key   string
+	Value string
+}
+
+// TopologyInfo denotes a mapping from TopologyPair to a string set
```
Comment (bsalamat) [inline pkg/scheduler/nodeinfo/topology_info.go]: The string set is a set of node names. Please state that in the comment. The current comment is not very helpful.

File: pkg/scheduler/internal/cache/cache.go [hunk 11]
```diff
@@ -63,6 +63,8 @@ type schedulerCache struct {
 	nodeTree  *NodeTree
 	// A map from image name to its imageState.
 	imageStates map[string]*imageState
+	// A map from node label to node names
+	toplogyInfo schedulernodeinfo.TopologyInfo
```
Comment (bsalamat) [inline pkg/scheduler/internal/cache/cache.go]: Do we really need to put this in the cache? Basically, we are building and maintaining this structure in the scheduler memory even if no pods in the cluster use inter-pod affinity. This increases scheduler's memory usage. Is there any strong reason for having it here?

**Review by bsalamat — COMMENTED** [review]
This PR is changed a lot after the last time I reviewed it. Here are my early comments. I still need to check the main part of the algorithm. My biggest concern so far is the introduction of a new structure in the cache which can use a lot of memory in larger clusters. Please see my comment below.

File: pkg/scheduler/nodeinfo/topology_info.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+	oldLabels, newLabels := oldNode.Labels, newNode.Labels
+	if helper.Semantic.DeepEqual(oldLabels, newLabels) {
+		return
+	}
+
+	toRemovePairs, toAddPairs := make([]TopologyPair, 0), make([]TopologyPair, 0)
+	for oldKey, oldVal := range oldLabels {
+		if newVal, ok := newLabels[oldKey]; ok {
+			if oldVal != newVal {
+				toAddPairs = append(toAddPairs, TopologyPair{oldKey, newVal})
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/nodeinfo/topology_info.go]: It was for readability, but yes, they're are unnecessary. Updated.

File: pkg/scheduler/algorithm/predicates/predicates.go [hunk 13]
```diff
@@ -1651,3 +1654,21 @@ func (c *VolumeBindingChecker) predicate(pod *v1.Pod, meta PredicateMetadata, no
 	klog.V(5).Infof("All PVCs found matches for pod %v/%v, node %q", pod.Namespace, pod.Name, node.Name)
 	return true, nil, nil
 }
+
+// BuildTopologyInfo buids a TopologyInfo based on a nodeInfoMap
+func BuildTopologyInfo(nodeInfoMap map[string]*schedulernodeinfo.NodeInfo) schedulernodeinfo.TopologyInfo {
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/predicates.go]: As tests in multiple package need it, it's now moved to the same package as `TopologyInfo`.

File: pkg/scheduler/algorithm/predicates/metadata.go [hunk 14]
```diff
@@ -215,6 +241,54 @@ func (topologyPairsMaps *topologyPairsMaps) appendMaps(toAppend *topologyPairsMa
...
+		}
+	}
+	return copy
+}
+
+func (aq affinityQuery) append(toAppend affinityQuery) {
+	for i := range toAppend {
+		if len(aq[i]) == 0 {
+			aq[i] = toAppend[i]
+			continue
```
[the remark this answers is not in the document]
Comment (Huang-Wei) [reply pkg/scheduler/algorithm/predicates/metadata.go]: Sometime I overused the "if negative path then continue" pattern :) Updated.

File: pkg/scheduler/nodeinfo/topology_info.go (hunk 9 above)
Comment (Huang-Wei) [reply pkg/scheduler/nodeinfo/topology_info.go]: I guess this file was initially written before refactoring. Moved to "internal" package.

File: pkg/scheduler/nodeinfo/topology_info.go (hunk 10 above)
Comment (Huang-Wei) [reply pkg/scheduler/nodeinfo/topology_info.go]: Done.

File: pkg/scheduler/internal/cache/cache.go (hunk 11 above)
Comment (Huang-Wei) [reply pkg/scheduler/internal/cache/cache.go]: The reason is that if we update the (global) topologyInfo only upon scheduling of an inter-affinity pod, it will cause an instant O(n) time. In current design, it's an average cost on each operation (update on node labels).

Comparing to the average time cost, the consistent space cost (memory footprint you pointed out) more concerns me. Let me see if it can be optimized.

**Comment by Huang-Wei** [comment]
/hold
to avoid accidental merge.

**Review by bsalamat — COMMENTED** [review]
I have attempted several times to review this PR, but unfortunately this PR has become so large and has made so many changes that were unnecessary to be in a single PR that it is very hard to follow and review.
I believe some of the refactoring and optimizations didn't need to be in the same PR. For example, introduction of `topology_info` to the cache and updating it at node event handlers didn't need to be a part of this PR. It could have been a follow up PR.
Generally, it is not a great idea to combine performance optimizations in the same PR that changes a major functionality.
